sm/agent/src/agent.rs

```diff
@@ -0,0 +1,326 @@
+use crate::{
+    error::AgentError,
+    state::{AgentStateRepository, State, StateRepository},
+};
+use json_sm::*;
+use log::{debug, error, info};
+use mqtt_client::{Client, Message, MqttClient, Topic, TopicFilter};
+use plugin_sm::plugin_manager::ExternalPlugins;
+use std::sync::Arc;
+use tedge_config::TEdgeConfigLocation;
+use tedge_users::{UserManager, ROOT_USER};
+
+#[derive(Debug)]
+pub struct SmAgentConfig {
+    pub request_topics: TopicFilter,
+    pub request_topic_list: Topic,
+    pub request_topic_update: Topic,
+    pub response_topic_list: Topic,
+    pub response_topic_update: Topic,
+    pub errors_topic: Topic,
+    pub mqtt_client_config: mqtt_client::Config,
+}
+
+impl Default for SmAgentConfig {
+    fn default() -> Self {
+        let request_topics =
+            TopicFilter::new("tedge/commands/req/software/#").expect("Invalid topic");
+
+        let request_topic_list =
+            Topic::new(SoftwareListRequest::topic_name()).expect("Invalid topic");
+
+        let request_topic_update =
+            Topic::new(SoftwareUpdateRequest::topic_name()).expect("Invalid topic");
+
+        let response_topic_list =
+            Topic::new(SoftwareListResponse::topic_name()).expect("Invalid topic");
+
+        let response_topic_update =
+            Topic::new(SoftwareUpdateResponse::topic_name()).expect("Invalid topic");
+
+        let errors_topic = Topic::new("tedge/errors").expect("Invalid topic");
+
+        let mqtt_client_config = mqtt_client::Config::default().with_packet_size(50 * 1024);
+
+        Self {
+            request_topics,
+            request_topic_list,
+            request_topic_update,
+            response_topic_list,
+            response_topic_update,
+            errors_topic,
+            mqtt_client_config,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SmAgent {
+    config: SmAgentConfig,
+    name: String,
+    user_manager: UserManager,
+    config_location: TEdgeConfigLocation,
+    persistance_store: AgentStateRepository,
+}
+
+impl SmAgent {
+    pub fn new(
+        name: &str,
+        user_manager: UserManager,
+        config_location: TEdgeConfigLocation,
+    ) -> Self {
+        let persistance_store = AgentStateRepository::new(&config_location);
+
+        Self {
+            config: SmAgentConfig::default(),
+            name: name.into(),
+            user_manager,
+            config_location,
+            persistance_store,
+        }
+    }
+
+    pub async fn start(&self) -> Result<(), AgentError> {
+        info!("Starting sm-agent");
```

Comment:
The names are not consistent.

* `sm-agent` here
* `tedge_agent` for the process
* `agent` for the crate directory